File: backend/app/api/audits.py
"""
Direct website audits - a single URL typed into the app, rather than an
uploaded CSV/XLSX. Reuses the exact same Job/Business pipeline as an import
(so crawling, checks, scoring, reporting, SSE progress and CSV/XLSX export
all work identically); the only difference is a one-row job created from a
URL instead of an uploaded file, and website discovery skips identity
matching since the user supplied the exact site to audit (see
discovery.verify_direct_website).
"""

from __future__ import annotations

import re
from typing import Any, Dict, Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from ..core.pipeline import manager
from ..core.urls import is_non_website_host, normalize_url, registrable_domain
from ..db import run_db
from ..events import activity
from ..models import Business, Job, JobItem
from ..settings import get_engine

router = APIRouter(prefix="/audits", tags=["audits"])


class QuickAudit(BaseModel):
    url: str
    name: Optional[str] = Field(default=None, max_length=255)
    start_immediately: bool = True


@router.post("/quick")
async def quick_audit(payload: QuickAudit) -> Dict[str, Any]:
    raw = (payload.url or "").strip()
    norm = normalize_url(raw)
    if not norm:
        raise HTTPException(
            status_code=400,
            detail="That does not look like a valid website URL. Include the domain, e.g. "
                   "example.com or https://example.com.",
        )
    is_profile, kind = is_non_website_host(norm)
    if is_profile:
        raise HTTPException(
            status_code=400,
            detail=f"That URL is a {kind.replace('_', ' ')}, not a standalone website, so it "
                   f"cannot be crawled and audited the same way.",
        )

    domain = registrable_domain(norm)
    label = (payload.name or "").strip() or domain or norm

    def _create(s) -> int:
        job = Job(
            name=f"Website audit — {label}"[:255],
            source_filename="", stored_path="",
            source_kind="url",
            original_columns=[], column_mapping={},
            engine_snapshot=get_engine(),
            status="queued", total=1,
        )
        s.add(job)
        s.flush()

        biz = Business(
            job_id=job.id, row_index=0, raw={"url": raw},
            name=label[:512], name_normalized=re.sub(r"[^a-z0-9]+", " ", label.lower()).strip(),
            website_original=norm,
            dedup_key=f"site:{domain}" if domain else "",
        )
        s.add(biz)
        s.flush()

        s.add(JobItem(job_id=job.id, business_id=biz.id, status="pending", stage="queued"))
        return job.id

    job_id = await run_db(_create)
    activity("", f"Direct audit started: {norm}", job_id=job_id, stage="job")

    started = False
    if payload.start_immediately:
        result = await manager.start(job_id)
        started = result.get("started", False)

    business_id = await run_db(
        lambda s: s.query(Business.id).filter(Business.job_id == job_id).scalar(), write=False
    )

    return {
        "job_id": job_id, "business_id": business_id, "url": norm, "started": started,
    }

File: cloudflare/src/api/audits.ts
// Direct website audits — POST /api/audits/quick. Ported from
// backend/app/api/audits.py. This is the only job-creation path the live
// frontend actually uses (see frontend/src/pages/Upload.jsx).
import { Hono } from "hono";
import type { Env } from "../types";
import { isNonWebsiteHost, normalizeUrl, registrableDomain } from "../core/urls";
import { getEngine } from "../lib/settings";
import * as q from "../db/queries";

export const audits = new Hono<{ Bindings: Env }>();

audits.post("/audits/quick", async (c) => {
  const body = await c.req.json().catch(() => ({}));
  const raw = String(body?.url || "").trim();
  const norm = normalizeUrl(raw);
  if (!norm) {
    return c.json({ detail: "That does not look like a valid website URL. Include the domain, e.g. example.com or https://example.com." }, 400);
  }
  const { isProfile, kind } = isNonWebsiteHost(norm);
  if (isProfile) {
    return c.json({ detail: `That URL is a ${kind.replace(/_/g, " ")}, not a standalone website, so it cannot be crawled and audited the same way.` }, 400);
  }

  const domain = registrableDomain(norm);
  const label = String(body?.name || "").trim() || domain || norm;
  const db = c.env.DB;
  const engine = await getEngine(db);

  const jobId = await q.createJob(db, { name: `Website audit — ${label}`.slice(0, 255), sourceKind: "url", total: 1, engineSnapshot: engine });
  const nameNormalized = label.toLowerCase().replace(/[^a-z0-9]+/g, " ").trim();
  const businessId = await q.createBusiness(db, {
    jobId, rowIndex: 0, raw: { url: raw }, name: label.slice(0, 512), nameNormalized,
    websiteOriginal: norm, dedupKey: domain ? `site:${domain}` : "",
  });
  await q.createJobItem(db, jobId, businessId);
  await q.addEvent(db, { jobId, message: `Direct audit started: ${norm}`, stage: "job" });

  const startImmediately = body?.start_immediately !== false;
  let started = false;
  if (startImmediately) {
    const instance = await c.env.AUDIT_WORKFLOW.create({ id: `job-${jobId}-biz-${businessId}`, params: { jobId, businessId } });
    await q.setJobWorkflowInstance(db, jobId, instance.id);
    started = true;
  }

  return c.json({ job_id: jobId, business_id: businessId, url: norm, started });
});
